### PR TITLE
Fix wildcard expansion when includes is empty and excludes has wildcards

### LINKED ISSUE #9566: No resources are backed up when excludedNamespaces field is specified and includedNamespace is nil
**What steps did you take and what happened:**
Create a backup specifying only excludedNamespaces field, velero didn't backup any resources and the backup was marked as completed.  Attaching details below:- 

```
$ cat backup.yaml
kind: Backup
apiVersion: velero.io/v1
metadata:
  name: test-backup
  namespace: openshift-adp
spec:
  excludedNamespaces:
  - 'openshift-*'

```

```
velero describe backup --details -n openshift-adp test-backup
Name:         test-backup
Namespace:    openshift-adp
Labels:       velero.io/storage-location=ts-dpa-1
Annotations:  velero.io/resource-timeout=10m0s
              velero.io/source-cluster-k8s-gitversion=v1.34.2
              velero.io/source-cluster-k8s-major-version=1
              velero.io/source-cluster-k8s-minor-version=34

Phase:  Completed


Namespaces:
  Included:  *
  Excluded:  openshift-*

Resources:
  Included cluster-scoped:    <none>
  Excluded cluster-scoped:    volumesnapshotcontents.snapshot.storage.k8s.io
  Included namespace-scoped:  *
  Excluded namespace-scoped:  volumesnapshots.snapshot.storage.k8s.io

Label selector:  <none>

Or label selector:  <none>

Storage Location:  ts-dpa-1

Velero-Native Snapshot PVs:    auto
File System Backup (Default):  false
Snapshot Move Data:            false
Data Mover:                    velero

TTL:  720h0m0s

CSISnapshotTimeout:    10m0s
ItemOperationTimeout:  4h0m0s

Hooks:  <none>

Backup Format Version:  1.1.0

Started:    2026-02-26 20:40:42 +0530 IST
Completed:  2026-02-26 20:40:44 +0530 IST

Expiration:  2026-03-28 20:40:42 +0530 IST

Total items to be backed up:  0
Items backed up:              0

Resource List:

Backup Volumes:
  Velero-Native Snapshots: <none included>

  CSI Snapshots: <none included>

  Pod Volume Backups: <none included>

HooksAttempted:  0
HooksFailed:     0
```

**What did you expect to happen:**

Velero should backup namespaces which are not matching to the excludedNamespace pattern. 

**The following information will help us better understand what's going on**:

_If you are using velero v1.7.0+:_  
Please use `velero debug  --backup <backupname> --restore <restorename>` to generate the support bundle, and attach to this issue, more options please refer to `velero debug --help` 

_If you are using earlier versions:_  
Please provide the output of the following commands (Pasting long output into a [GitHub gist](https://gist.github.com) or other pastebin is fine.)
- `kubectl logs deployment/velero -n velero`
- `velero backup describe <backupname>` or `kubectl get backup/<backupname> -n velero -o yaml`
- `velero backup logs <backupname>`
- `velero restore describe <restorename>` or `kubectl get restore/<restorename> -n velero -o yaml`
- `velero restore logs <restorename>`

**Anything else you would like to add:**

**Environment:**

- Velero version (use `velero version`):  Velero 1.18
- Velero features (use `velero client config get features`): 
- Kubernetes version (use `kubectl version`):
- Kubernetes installer & version:
- Cloud provider or hardware configuration:
- OS (e.g. from `/etc/os-release`):

**Vote on this issue!**

This is an invitation to the Velero community to vote on issues, you can see the project's [top voted issues listed here](https://github.com/vmware-tanzu/velero/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).  
Use the "reaction smiley face" up to the right of this comment to vote.

- :+1: for "I would like to see this bug fixed as soon as possible"
- :-1: for "There are more important bugs to focus on right now"

### PR DESCRIPTION
**Pull request by Joeavaikath**
## Summary

Fixes #9566

When a Backup CR is created via `kubectl apply` with empty `includedNamespaces`
and a wildcard pattern in `excludedNamespaces` (e.g., `test-backup*`),
`ShouldExpandWildcards` triggers expansion. The empty includes expands to nil,
but `wildcardExpanded` is set to true, causing `ShouldInclude` to return false
for all namespaces — effectively backing up nothing.

This fix normalizes empty `IncludedNamespaces` to `["*"]` in the backup
controller's `prepareBackupRequest`, before any downstream processing. This:
1. Prevents the `wildcardExpanded + empty includes` guard from rejecting everything
2. Matches the CLI path's behavior where includes defaults to `["*"]`
3. Ensures consistent Namespace CR inclusion — excluded namespaces' Namespace
   objects are still backed up, matching Velero's intentional behavior
   (see `item_backupper.go` `itemInclusionChecks`)

## Changes
- **`pkg/controller/backup_controller.go`**: In `prepareBackupRequest()`, normalize empty `IncludedNamespaces` to `["*"]` before namespace validation
- **`pkg/controller/backup_controller_test.go`**: Add test verifying empty includes is normalized to `["*"]`

## Test plan
- [ ] Existing tests pass (`go test ./pkg/controller/ ./pkg/util/collections/ ./pkg/util/wildcard/`)
- [ ] New test: `TestPrepareBackupRequest_EmptyIncludedNamespacesNormalizedToWildcard` verifies normalization
- [ ] Manual: `kubectl apply` a Backup CR with empty `includedNamespaces` and wildcard `excludedNamespaces` — verify correct namespaces are backed up
- [ ] 

# Please indicate you've done the following:

- [x] [Accepted the DCO](https://velero.io/docs/v1.5/code-standards/#dco-sign-off). Commits without the DCO will delay acceptance.
- [x] [Created a changelog file (`make new-changelog`)](https://velero.io/docs/main/code-standards/#adding-a-changelog) or comment `/kind changelog-not-required` on this PR.
- [ ] Updated the corresponding documentation in `site/content/docs/main`.- [ ]

### CODE REVIEW

File: pkg/util/collections/includes_excludes.go [hunk 1]
```diff
@@ -164,6 +164,16 @@ func (nie *NamespaceIncludesExcludes) ExpandIncludesExcludes() error {
...
 		}
 
+		// Empty includes means "include all", so normalize to "*" to
+		// prevent the wildcardExpanded+empty guard from excluding
+		// everything, and to match the CLI path's behavior.
+		// Note: Namespace CRs for excluded namespaces will still be
+		// backed up — this is intentional Velero behavior (see
+		// item_backupper.go itemInclusionChecks).
+		if len(includes) == 0 {
+			expandedIncludes = []string{"*"}
```
Comment (blackpiglet) [inline pkg/util/collections/includes_excludes.go]: The change itself is sensible, but it appears to be more relevant to setting the default value for the backup, rather than the include/exclude filter setting.

I recommend relocating this fix to the backup controller code. The code in that location is better equipped to manage and normalize the backup Custom Resource (CR).


https://github.com/vmware-tanzu/velero/blob/f854a0653abcf785f679073ec1f8cb63d036a8ef/pkg/controller/backup_controller.go#L578-L584

Comment (Joeavaikath) [reply pkg/util/collections/includes_excludes.go]: That sounds better, will make this change

**Review by blackpiglet — DISMISSED** [review]

File: pkg/controller/backup_controller.go [hunk 2]
```diff
@@ -570,6 +570,13 @@ func (b *backupReconciler) prepareBackupRequest(ctx context.Context, backup *vel
 		}
 	}
 
+	// Empty IncludedNamespaces means "include all namespaces". Normalize
+	// to ["*"] so that downstream wildcard expansion does not collapse
+	// an empty-includes + wildcard-excludes combination into "back up nothing".
```
Comment (shubham-pampattiwar) [inline pkg/controller/backup_controller.go:575]: The root cause is in ShouldExpandWildcards, it triggers expansion when includes is empty and only excludes have wildcards. But there's nothing to expand in an empty includes list, and globStringSet.match already handles wildcard excludes natively. Fixing it there would protect all callers, not just the backup controller path. Not a blocker, just think it's a more targeted fix.

File: pkg/controller/backup_controller_test.go [hunk 3]
```diff
@@ -320,6 +320,34 @@ func TestBackupLocationLabel(t *testing.T) {
 	}
 }
 
+func TestPrepareBackupRequest_EmptyIncludedNamespacesNormalizedToWildcard(t *testing.T) {
```
Comment (shubham-pampattiwar) [inline pkg/controller/backup_controller_test.go:323]: The test verifies normalization happens but doesn't exercise the actual broken scenario, empty includes + wildcard excludes through NamespaceIncludesExcludes. That would be stronger regression coverage.

File: pkg/controller/backup_controller.go (hunk 2 above)
Comment (blackpiglet) [reply pkg/controller/backup_controller.go:575]: If so, I think we can have the fix in both places.

**Comment by Lyndon-Li** [comment]
@Joeavaikath Could you check if all the comments have been addressed so this PR could be merged? This is a task of 1.18.1 which will be RC very soon.
Please also help to make a cherry-pick after it is merged.

File: pkg/controller/backup_controller_test.go (hunk 3 above)
Comment (Joeavaikath) [reply pkg/controller/backup_controller_test.go:323]: Added a test to ShouldExpandWildcards

File: pkg/controller/backup_controller.go (hunk 2 above)
Comment (Joeavaikath) [reply pkg/controller/backup_controller.go:575]: Added a check to bypass expansion in such a case, even though normalization means this wouldn't occur
Better to have that check here as well

**Review by shubham-pampattiwar — APPROVED** [review]

File: pkg/util/wildcard/expand.go [hunk 4]
```diff
@@ -9,6 +9,11 @@ import (
 )
 
 func ShouldExpandWildcards(includes []string, excludes []string) bool {
+	// Empty includes is equivalent to * (match all) - don't expand
+	if len(includes) == 0 {
+		return false
```
Comment (blackpiglet) [inline pkg/util/wildcard/expand.go:14]: The fix is reasonable for the main branch now, but release-1.18 expands asterisk.
We should consider that and make modifications accordingly for release-1.18.

**Review by blackpiglet — APPROVED** [review]

File: pkg/util/wildcard/expand.go (hunk 4 above)
Comment (Joeavaikath) [reply pkg/util/wildcard/expand.go:14]: Will backport to release-1.18 branch
